---
layout: default
title: "Research"
description: "Research projects in the lab"
group: navigation
navorder: 1
---
{% include JB/setup %}
<div class="bigspacer"></div>
<div class="construction">
	<hr/>
Research Themes
  </div>
  <font size="3" color="black">Research in the Bik Lab is intensely interdisciplinary, using high-throughput sequencing and diverse –Omics approaches to explore broad patterns in microbial eukaryote assemblages (biodiversity and phylogeography, functional roles for microbial taxa, and the relationship between species and environmental parameters), with an emphasis on nematodes in marine sediments. Our long-term research interests lie at the interface between biology and computer science, using biological questions and evolutionary hypotheses to drive the development and refinement –Omic approaches focused on marine microbial eukaryotes. Microbial eukaryotes (organisms <1mm, such as nematodes, fungi, protists, and other ‘minor’ metazoan phyla) are abundant and ubiquitous across every ecosystem on earth, performing key functions such as nutrient cycling and sediment stability in marine habitats. Yet, their unexplored diversity represents one of the major challenges in biology and currently limits our capacity to understand, mitigate and remediate the consequences of environmental change. Given this knowledge gap, my research vision spans three key areas: 
<p></p>
<b>1) "Is Everything Everywhere?": Testing seminal biodiversity hypotheses using –Omics</b>
Marine ecosystems have been characterized by a number of overarching, large-scale patterns, including extremely high biodiversity, peaks of species diversity at intermediate depths, and latitudinal gradients of diversity (e.g. decreasing towards the poles; <a href="http://www.biogeosciences.net/7/2851/2010/bg-7-2851-2010.html">Ramirez-Llodra et al. 2010, Biogeosciences, 7: 2851</a>). To date, these patterns have been derived from work based on larger organisms (macrofauna and megafauna) and studies using non-molecular classical methods (e.g. morphological taxonomy in studies of microbial metazoa). One of my primary research aims is to revisit and reassess these reported biogeographic patterns for sediment-dwelling organisms, by amassing large volumes of molecular data that will be compared to the vast body of existing morphological studies (e.g. putatively “cosmopolitan” morphospecies). My work will go far beyond previous studies by investigating the full breadth of microbial taxa in marine sediments (e.g. microbial eukaryotes, fungi, bacteria, archaea, viruses). Using my existing collection of frozen sediment cores (>500 sites) and large network of deep-sea collaborators to collect future samples, my work will leverage a truly global sample set that is representative of diverse geographic locations (North Atlantic, Eastern Pacific, Arctic, Antarctic, Gulf of Mexico, etc.), types of sediment habitats (e.g. abyssal plains, canyons, cold seeps, whale falls), and water depths (200-5000+m). These methods will allow us to address the questions: do molecular data and deep sequencing approaches agree with known patterns of diversity and biogeography in the marine sediment habitats? Does biogeography and alpha/beta species diversity vary across taxonomic groups (prokaryotes, fungi, protists, and microbial metazoa)?
  
<b>2) “Systems Ecology” to investigate ecosystem function and species interactions, with an emphasis on microbial eukaryotes in marine sediments</b>
Inspired by systems biology, a “systems ecology” approach aims to provide a more unrestricted view of natural ecosystems, by using large data volumes to enable the discovery and description of complex interactions within biological systems. Although my research focuses on microbial eukaryotes (with an emphasis on metazoan groups such as nematodes), the nature of “systems ecology” enables exploration and analysis of a much wider taxonomic breadth. For example, shotgun metagenomics will inherently sequence DNA from bacteria, archaea, and viruses, quantifying functional potential and novel diversity in these groups, and in some cases even allowing for computational reconstruction of small genomes. By using such a systems ecology approach, my research will go beyond “Is Everything Everywhere?”, by also asking “are functional pathways (genes, metabolites) and host-microbe interactions conserved in similar habitats?” and “does taxonomic diversity correspond with functional diversity?”. A systems ecology approach is increasingly relevant because of the replenishing “microbial seed bank” that appears to persist in the oceans (<a href="http://www.nature.com/ismej/journal/v6/n6/full/ismej2011162a.html">Caporaso et al. 2012, ISME, 6:1089</a>) and mounting evidence that functional genes are a more important factor for community assembly versus species themselves (<a href="http://www.pnas.org/content/108/34/14288.abstract">Burke et al. 2011, PNAS, 108:14288</a>). In order to move towards “systems ecology,” we must move beyond environmental rRNA surveys, incorporating other types of –Omic data to better quantify community assemblages, metabolic pathways, and putative ecological roles. The ultimate goal is to gain a holistic view of ecosystem function. However, making this leap will require new computational tools. Metagenomes currently represent one of the most complex type of environmental datasets we are able to generate, and “taxonomic silos” still persist for the analysis of environmental data; computational analyses of bacteria/archaea, microbial eukaryotes, and viruses must currently be carried out using separate databases and independent software tools. Thus, advancing the bioinformatics toolkit is inherently intertwined with my hypotheses-driven research vision. One goal is to foster the development of increasingly flexible and inclusive bioinformatics pipelines, to enable rapid, integrative analyses of all microbial taxa across independent –Omics datasets. Such integrated pipelines—especially those harnessing phylogeny-driven algorithms—are key for “systems ecology.” A second goal is to develop intuitive, exploratory data visualization frameworks as a new paradigm for the analysis of large –Omic datasets spanning multiple taxonomic domains. 
  
<b>3) Comparative phylogenomics and targeted genome sequencing of microbial eukaryotes as a complement to environmental –Omics</b>
Physical isolation and sequencing of individual microbial eukaryote specimens is now possible, via single-cell approaches and/or protocols designed for low amounts of DNA. Using targeted genome sequencing of individual specimens from marine sediments (focusing on marine metazoan species such as nematodes, tardigrades, kinorhynchs, and other ‘minor’ phyla), my research will incorporate a large-scale effort to fill in the branches of the eukaryotic tree of life. This final facet of my research vision aims to address questions such as: Are there divergent evolutionary lineages and endemic hot spots in unique or isolated seafloor habitats? Can we reconstruct the historical evolutionary trajectory of the deep sea (as a source vs. sink habitat)? Current eukaryotic databases are exceedingly sparse (compared to known species diversity)—this paucity of data critically limits our ability to analyze and interpret environmental –Omic datasets or reconstruct deep phylogeny. In particular, the interpretation of metagenomic data is still inherently reliant on comparisons to available genome sequences and their corresponding annotations; the taxonomic utility of phylogenetic marker genes is also constrained by the coverage of the original reference database used to build the markers. A robust, openly accessible collection of marine metazoan species in public databases will represent a significant long-term community resource that can be re-analyzed and mined in future environmental studies, and also used to inform predictive modeling approaches. Furthermore, genome sequences will allow us to link environmental contigs with a known taxonomic ID (e.g. “hypothetical” proteins), identify clade-specific marker genes for taxonomic classification and phylogenetics, and identify the genomic repertoire (thus inferring putative ecological functions) of each eukaryotic specimen.

  	</font>
  	<hr/>

{% for project in site.categories.projects %}
<div class='row'>
  <!--
  <div class='row'>
    <div class='col-md-12'>
    <div class="media">
      <a class="project-image" href="{{ project.url }}">
        <img class="media-object" src="{{ project.image }}">
      </a>
    </div>
  </div>
  -->
  <div class='row'>
    <div class="col-md-4 paperbox">
   <div class="smallhead media-heading"><a href="{{ project.url }}" class="off"><img width=300 class="media-object pull-right" src="{{ project.image }}"/></a></div>
    </div>
    <div class='col-md-8'>
      {{ project.content }}
    </div>
  </div>
</div>
{% endfor %}

<div class="bigspacer"></div>
