---
layout: project
title: "Collaborations"
tagline: "Collaborative Projects"
handle: phylosift
image: /assets/images/projects/phylosift-logo.png
category: project
tags: [phylosift, software, collaborations]
---
{% include JB/setup %}

## PhyloSift (External Collaboration)

PhyloSift is a software pipeline for phylogenetic analysis of genomes and metagenomes. Using any biological sequence as input data (nucleotide or amino acid), PhyloSift uses a reference database of profile HMMs to identify candidate sequences matching phylogenetically-informative marker genes. Candidate sequences identified from input data are then subjected to phylogenetic placement approaches, where short reads are inserted into reference phylogenies and given taxonomic assignments based on this tree placement. PI Holly Bik was actively involved in development and documentation of the PhyloSift pipeline, and we continue to use this software for ongoing metagenomics data analysis. Software download and extensive documentation are available on the [main PhyloSift website], and the code is maintained as an [open source repository on Github].

<u>Reference:</u> Darling A, Jospin G, Lowe E, Matsen FA, Bik HM, Eisen JA (2014) [PhyloSift: phylogenetic analysis of genomes and metagenomes], <em>PeerJ</em>, 2:e243.


[main PhyloSift website]: https://phylosift.wordpress.com/
[open source repository on Github]: https://github.com/gjospin/PhyloSift 
[PhyloSift: phylogenetic analysis of genomes and metagenomes]: http://dx.doi.org/10.7717/peerj.243