---
layout: paper
title: "Microbiota of the indoor environment: a meta-analysis"
year: "2015"
shortref: "Adams <i>et al.</i> 2015 <i>Microbiome</i>"
nickname: 
journal: "Microbiome"
volume: 3
issue: 
pages: 49
authors: "Adams RI, Bateman AC, <b><u>Bik HM</u></b>, Meadow JF"
image: /assets/images/papers/default-paper.svg
pdf: /assets/pdfs/Adams-2015-Microbiome-IndoorEnviron.pdf
pdflink:
github: 
<!-- pmid: 15644440
pmcid: PMC545518
f1000: http://f1000.com/prime/1030269
figshare:  -->
doi: 10.1186/s40168-015-0108-3
category: paper
published: true
peerreview: true
review: false
tags: 
---
{% include JB/setup %}

# Abstract

<u>Background:</u> As modern humans, we spend the majority of our time in indoor environments. Consequently, environmental exposure to microorganisms has important implications for human health, and a better understanding of the ecological drivers and processes that impact indoor microbial assemblages will be key for expanding our knowledge of the built environment. In the present investigation, we combined recent studies examining the microbiota of the built environment in order to identify unifying community patterns and the relative importance of indoor environmental factors. Ultimately, the present meta-analysis focused on studies of bacteria and archaea due to the limited number of high-throughput fungal studies from the indoor environment. We combined 16S ribosomal RNA (rRNA) gene datasets from 16 surveys of indoor environments conducted worldwide, additionally including 7 other studies representing putative environmental sources of microbial taxa (outdoor air, soil, and the human body).

<u>Results:</u> Combined analysis of subsets of studies that shared specific experimental protocols or indoor habitats revealed community patterns indicative of consistent source environments and environmental filtering. Additionally, we were able to identify several consistent sources for indoor microorganisms, particularly outdoor air and skin, mirroring what has been shown in individual studies. Technical variation across studies had a strong effect on comparisons of microbial community assemblages, with differences in experimental protocols limiting our ability to extensively explore the importance of, for example, sampling locality, building function and use, or environmental substrate in structuring indoor microbial communities.

<u>Conclusions:</u> We present a snapshot of an important scientific field in its early stages, where studies have tended to focus on heavy sampling in a few geographic areas. From the practical perspective, this endeavor reinforces the importance of negative “kit” controls in microbiome studies. From the perspective of understanding mechanistic processes in the built environment, this meta-analysis confirms that broad factors, such as geography and building type, structure indoor microbes. However, this exercise suggests that individual studies with common sampling techniques may be more appropriate to explore the relative importance of subtle indoor environmental factors on the indoor microbiome.