---
layout: paper
title: "HybPiper: Extracting Coding Sequence and Introns for Phylogenetics from High- Throughput Sequencing Reads Using Target Enrichment"
image: /assets/images/papers/HybPiper_Infographic.jpg
authors: <b><u>Johnson MG</b></u>, Gardner EG, Liu Y, Medina R, Goffinet B, Shaw AJ, Zerega NJC, and Wickett NJ
year: 2016
shortref: Johnson et al. 2016. APPS.
journal: "Applications in Plant Sciences"
doi: 10.3732/apps.1600016
github: https://github.com/mossmatters/HybPiper
pdf: /assets/pdfs/papers/johnson-hybpiper.pdf
volume:
issue:
pages:
category: paper
published: true
peerreview: true
review: false

---

# Abstract

*Premise of the study*: Using sequence data generated via target enrichment for phylogenetics requires reassembly of high- throughput sequence reads into loci, presenting a number of bioinformatics challenges. We developed HybPiper as a user- friendly platform for assembly of gene regions, extraction of exon and intron sequences, and identi cation of paralogous gene copies. We test HybPiper using baits designed to target 333 phylogenetic markers and 125 genes of functional signi cance in Artocarpus (Moraceae).
*Methods and Results*: HybPiper implements parallel execution of sequence assembly in three phases: read mapping, contig as- sembly, and target sequence extraction. The pipeline was able to recover nearly complete gene sequences for all genes in 22 species of Artocarpus. HybPiper also recovered more than 500 bp of nontargeted intron sequence in over half of the phylogenetic markers and identi ed paralogous gene copies in Artocarpus.*Conclusions*: HybPiper was designed for Linux and Mac OS X and is freely available at [https://github.com/mossmatters/HybPiper](https://github.com/mossmatters/HybPiper).
