---
layout: paper
title: "Allele phasing has minimal impact on phylogenetic reconstruction from targeted nuclear gene sequences in a case study of Artocarpus"
image: /assets/images/papers/allele-phasing.jpg
authors: HR Kates, <b><u>Johnson MG</b></u>, Gardner EM, Zerega NJC, and Wickett N.
year: 2018
shortref: Kates et al. 2018. APPS.
journal: "American Journal of Botany"
doi: 10.1002/ajb2.1068
pdf: /assets/pdfs/papers/Kates-allelephasing.pdf
github: https://github.com/mossmatters/phyloscripts
volume: 105
issue: 3
pages: 404-416
category: paper
published: true
peerreview: true
review: false

---

# Abstract


*Premise of the Study*

Untapped information about allele diversity within populations and individuals (i.e., heterozygosity) could improve phylogenetic resolution and accuracy. Many phylogenetic reconstructions ignore heterozygosity because it is difficult to assemble allele sequences and combine allele data across unlinked loci, and it is unclear how reconstruction methods accommodate variable sequences. We review the common methods of including heterozygosity in phylogenetic studies and present a novel method for assembling allele sequences from target‐enriched Illumina sequencing libraries.

*Methods*

We performed supermatrix phylogeny reconstruction and species tree estimation of *Artocarpus* based on three methods of accounting for heterozygous sequences: a consensus method based on de novo sequence assembly, the use of ambiguity characters, and a novel method for incorporating read information to phase alleles. We characterize the extent to which highly heterozygous sequences impeded phylogeny reconstruction and determine whether the use of allele sequences improves phylogenetic resolution or decreases topological uncertainty.

*Key Results*

We show here that it is possible to infer phased alleles from target‐enriched Illumina libraries. We find that highly heterozygous sequences do not contribute disproportionately to poor phylogenetic resolution and that the use of allele sequences for phylogeny reconstruction does not have a clear effect on phylogenetic resolution or topological consistency.

*Conclusions*

We provide a framework for inferring phased alleles from target enrichment data and for assessing the contribution of allelic diversity to phylogenetic reconstruction. In our data set, the impact of allele phasing on phylogeny is minimal compared to the impact of using phylogenetic reconstruction methods that account for gene tree incongruence.

